DBZ-438 Avoiding issues due to concurrent usage of BufferedBlockingCo…

gunnarmorling opened the pull request:
…nsumer

jpechane commented on debezium-core/src/main/java/io/debezium/function/BufferedBlockingConsumer.java:
```diff
@@ -42,27 +43,31 @@ public default void flush() throws InterruptedException {
             public void accept(T t) throws InterruptedException {
-                if (last != null) delegate.accept(last);
-                last = t;
+                T previous = last.getAndSet(t);
```
This code changes the order of execution in comparison to the previous code thus I opted for `synchronized`. I am not sure if the order matters or not

rhauch replied:
Isn't it only changing the order of execution to be an atomic swap?

gunnarmorling replied:
> What is different about this approach is that it does not prevent the accept and flush from being called concurrently.

Right, it was my thinking that the calling code needs to make sure that it only calls `flush()` after the `accept()` calls.

jpechane commented on debezium-core/src/main/java/io/debezium/function/BufferedBlockingConsumer.java:
```diff
@@ -42,27 +43,31 @@ public default void flush() throws InterruptedException {
+                        delegate.accept(function.apply(previous));
                     } finally {
-                        last = null;
+                        last.set(null);
```
Is it possible to run `accept` and `flush` in parallel? If yes then this code will overwrite the value set in `accept` when order of execution is unfortunate.

rhauch replied:
Yeah, it may be possible unless we explicitly change the contract and make sure MongoDB isn't doing this. 
But what if this method were to do the following:

```
T previous = last.getAndSet(null);
if (previous != null) {
    delegate.accept(function.apply(previous));
}
```

In the single-threaded case, there is obviously no concurrency. If multiple threads are concurrently calling `accept(T)`, then this class would work except when one thread get preempted part-way through and another thread calls `accept(T)` and reorders the responses. The same is true if `accept(T)` and `flush(...)` are called concurrently.

However, I'm not sure this matters in the multi-threaded case, right? If there are multiple threads, how can you expect to maintain the order of records from those threads? The records from any single thread will be forwarded in the same order, and IMO that's about as much as we can expect.

And, if you're using this with multiple threads, it seems like it's the responsibility of the caller(s) to ensure that `flush(...)` is called *_after_* all of the `accept(T)` methods have been called. Seems like we need to make sure that MongoDB is doing that, and it might be worth documenting this a bit more.

gunnarmorling replied:
> And, if you're using this with multiple threads, it seems like it's the responsibility of the caller(s) to ensure that flush(...) is called after all of the accept(T) methods have been called. 

Yes, that was my thinking as well and I'd say it is like this in the current MongoDB connector code. `flush()` is only called after the latch is passed, i.e. all the threads have been doing their work.

gunnarmorling replied:
I wasn't concerned about ordering of messages emitted by different threads, and I don't think `synchronized` vs. this proposal does make a difference in this regard. My concern was about the synchronization of all the `accept()` calls by all the threads. I.e. I'm afraid parallelism of these threads is largely hindered by forcing them through this synchronization. That's what I seeked to avoid by using the atomic reference.

jpechane replied:
So, let's document the limitation on `flush` - soething like - this method can be executed if and only if all calls to `accept` has been completed and we can close the issue.

gunnarmorling replied:
Done.

rhauch commented:
I might have a different suggestion. I actually think the synchronized approach is much simpler, but I agree that in the single-threaded case it's not ideal and has too much overhead. How about having the existing `bufferLast(...)` method creating an unsynchronized implementation (as-is), and adding another `concurrentBufferLast(...)` method (perhaps with a better name?) that creates a synchronized implementation? Then, the MongoDB connector uses the concurrent one.

jpechane commented on debezium-core/src/main/java/io/debezium/function/BufferedBlockingConsumer.java:
```diff
@@ -42,27 +43,31 @@ public default void flush() throws InterruptedException {
             @Override
-            public void flush(Function<T, T> function) throws InterruptedException {
-                if (last != null) {
+            public synchronized void flush(Function<T, T> function) throws InterruptedException {
```
Is the method `synchronized intentionally`?

gunnarmorling replied:
Yes, I thought it'd be good to prevent concurrent `flush()` calls.

jpechane commented:
Two implementations is definitely tempting idea. And wrt `synchronized` vs atomic - I ma not sure we are focusing on the right optimization.  Will atomic variable really remove thread contention or only reduce it? If it is the latter case then we should seek for a way how to remove it completely.

jpechane commented:
I ran a test that was disclosing the issue and this code passed.

gunnarmorling commented:
I've rebased this and force-pushed an update. I've renamed `flush()` to `close()`, making it more clear that it only should be called once at the end. That way it should be fine to work with the reference and avoid contention in `accept()` and it can be safely closed afterwards.

jpechane commented:
Fine for me, thanks!
